[ADR] 25: RPC Streaming (#1431)
Azure/iot-operations-sdks: doc/dev/adr/0025-rpc-streaming.md

 - The protocol MUST support multiple independent streaming commands simultaneously.
 - The invoker and the executor MUST each be able to cancel the streamed request and/or the streamed response at any time.
 - The invoker and the executor MUST be able to send their requests and responses at arbitrary times relative to one another, except that the invoker MUST initiate the streaming invocation with a request.
   - For instance, the executor MAY send a response as soon as it receives a request, or it MAY wait until the request stream ends before sending its first response.
   - Likewise, the invoker MAY send a request in reaction to a received response.
 - The invoker and the executor MUST each be able to end their own stream gracefully at any time.
   - A side that does not know in advance whether an entry will be its last MUST still be able to end its stream afterward without sending another full entry.

## Non-requirements

 - Different payload shapes per command response/request
 - Chunking a single large payload across multiple stream messages
   - Any stream message may be lost (message expiry or other circumstances), so a chunked payload could not be reliably reassembled.

>  - The API of the receiving side of a stream will provide the user the streamed requests/responses in their **intended** order rather than their **received** order

Reviewer: Not really sure what this means, or how that would be possible given the caveats listed below.

How would you know the nth message was lost due to broker expiry, and that it's okay to deliver the n+1th message when the protocol is designed to provide the messages in the **intended** order? Trying to deliver in an intended order also will force some kind of received message caching, which would then need an arbitrary cache clean timeout, etc.

I find this confusing

Reviewer 2: edit: Did not notice this was in the **non-requirements** section

Each entry in a request or response stream pairs a user **payload** with **metadata**. That metadata combines two scopes:

- **Message metadata** is scoped to the individual stream entry it travels with.
- **Stream metadata** applies to a whole stream. The request and response streams carry **different** stream metadata, so this scope is **asymmetric** across directions. On the wire it repeats on every message to survive first-message loss, and is read once.

#### Core abstractions

Each side both produces and consumes a stream — the invoker produces requests and consumes responses, the executor does the reverse — and the two together form one **exchange**, the unit of cancellation and timeout.

The streams themselves surface as plain async sequences — you **supply** the one you produce and **iterate** the one you consume — so producing and consuming take different shapes. Completion, cancellation, and timeout are **exchange-scoped** — one per invocation, not per direction — so a single cancel or timeout covers the whole exchange rather than either individual stream (see [cancellation](#cancellation-support) and [timeout](#timeout-support)).

#### Exchange completion

> Each producer ends its own stream with an **`isLast`** signal. An exchange is **gracefully complete** only when *both* of its streams have closed: the invoker has sent its `isLast` request **and** received the `isLast` response, and symmetrically the executor has received the `isLast` request **and** sent the `isLast` response. Closing one stream (via `isLast`) does **not** end the exchange — a side that finishes its own stream early stays active until the other stream closes too, or until the [exchange timeout](#exchange-level-timeout) fires. Any other terminal — cancellation or timeout — ends the whole exchange immediately. Requiring both streams to close is the shared definition of completion used by [timeout](#timeout-support) and [cancellation](#cancellation-support).

Reviewer: nitpick, but i would prefer the name `terminal` or `sentinel` or even `last` to refer to this rather than `isLast`.

- ```c:4:last:6``` — the request stream's final (`isLast`) message at index `4`; the invoker's remaining exchange timeout has dropped to 6 seconds (the field is a live countdown).
- ```c:2:cancel:10``` — a request-direction cancellation at the producer's next index (`2`).
- ```s:1:10``` — a request-direction status about the received message at stream index `1`; its `__stat` value carries the error status for that message.
- ```d:0``` / ```c:7:last``` / ```s:3``` — the response-direction counterparts (executor → invoker); identical forms except `timeout_length` is omitted.

Every MQTT PUBLISH belonging to a streaming exchange must include `__stream` in exactly one of these three forms: data messages use the `d` form, control messages use the `c` form (`c:…:cancel`, `c:…:last`), and status messages use the `s` form. A status message carries no stream entry; its `__stat` property carries the outcome details for the received message it references.

[see cancellation support](#cancellation-support) and [timeout support](#timeout-support) for how these fields are used.

#### Topics and routing

> A single **correlation GUID** identifies the whole exchange; every message of both streams carries it. The exchange uses **two MQTT topics** — the **command topic** (invoker → executor) and the **response topic** (executor → invoker) — and each carries that direction's data, control, and status messages together.

Reviewer 3: ```suggestion
A single **correlation GUID** identifies the whole exchange; every message of both streams carries it. The exchange uses **two MQTT topics** — the **request topic** (invoker → executor) and the **response topic** (executor → invoker) — and each carries that direction's data, control, and status messages together.
```
unless there was a reason to use command topic instead of our usual request topic semantics from vanilla rpc

The executor subscribes to the **command topic** with a **shared subscription** (so that, with multiple executors, only one handles each exchange), with the same topic pre/suffixing and custom-topic-token support as vanilla RPC. The **response topic** is `clients/{invoker client id}/...` (prefixed like vanilla RPC), unique to the invoker and not shared; the invoker subscribes to it before publishing. Each request-stream message the invoker publishes carries the response topic (so the executor knows where to reply) and a `$partition` user property set to the invoker's client id.

Because the command topic is a shared subscription, **every** command-topic packet for an exchange — request data, an `isLast` request, invoker cancellation, and the invoker's `Canceled` acknowledgement — must carry that same `$partition`, so the broker routes them all to the same executor; otherwise a control packet could reach a different executor that holds no state for the correlation and be silently dropped. Response-topic packets need only the correlation data, since that topic is already unique to the invoker.

#### Exchange lifetime

> Because closing one stream does not end the exchange (see [exchange completion](#exchange-completion)), each endpoint keeps its per-correlation state active until the exchange is terminal, so control still flows after an `isLast`. Once a side is terminal, further data messages for that correlation are acknowledged and ignored; only the required control re-answers (for example, re-sending `Canceled` for a re-issued cancellation) are sent. The per-correlation state is kept as a **tombstone** so late or duplicate packets remain routable and are not treated as a new stream; it is retained at least as long as the longest expiry of any packet that could still arrive.

Reviewer 3: right? Just to be safe
```suggestion
Because closing one stream does not end the exchange (see [exchange completion](#exchange-completion)), each endpoint keeps its per-correlation state active until the exchange is terminal, so control still flows after an `isLast`. Once a side is terminal, further data messages for that correlation are acknowledged and ignored; only the required control re-answers (for example, re-sending `Canceled` for a re-issued cancellation) are sent. The per-correlation state is kept as a **tombstone** so late or duplicate packets remain routable and are not treated as a new stream; it is retained at least as long as the longest expiry of any packet that could still arrive, plus some buffer.
```

#### Producing and consuming

A side **consumes** one stream and **produces** the other. The consuming and producing rules below are identical for both roles; two further rules apply to the **executor** only.

**Consuming a stream:**

- **De-dup caching.** A consumer de-dups received data messages (QoS 1 may re-deliver) by correlationId + index — the index distinguishes duplicates since the correlationId is shared by the whole stream. Each cache entry is retained for the duration of its message's expiry interval (see [message level timeout](#message-level-timeout)), even beyond the end of the stream: clearing it when the stream finishes would let a late re-delivery still within its expiry window be treated as new, which is unsafe for non-idempotent commands.

> - **Acknowledgement.** By default a consumer acknowledges each message as soon as it is delivered to the user. Manual acknowledgement — finishing processing before forgoing broker re-delivery on an unexpected crash — is available only to the **executor** (request consumption); the invoker cannot resume a response stream after a crash, so it always auto-acknowledges.

Reviewer: This is not really how RPC works in Rust at least, so this would not follow the design that we have for the existing protocol, although perhaps that is inevitable given the streaming requirements.

The way we've done RPC is that the ack is not sent on the request PUBLISH until after the response PUBLISH is sent. The user is blind to any MQTT acking, they only interact with protocol level requests/responses. If they fail to respond and drop the mechanism for doing so, an automatic 500 response is sent in their stead.

I dont love the idea of exposing a switch related to the underlying MQTT protocol, so if we have to have this behavior i'd like it to be framed in the context of the RPC stream to blackbox the implementation.

Reviewer 2: We could call this a "stream entry ack" to make it an RPC streaming protocol-level construct rather than an MQTT level construct.

- The **invoker** starts its timer when it sends its first request. If it elapses before [graceful completion](#exchange-completion), it reports the timeout to the user and stops sending.
- The **executor** starts its timer on the first request it receives. If it elapses before [graceful completion](#exchange-completion), it reports the timeout to the user.

Every **request-direction** message (invoker → executor) carries the **current invoker timeout counter value** in the `timeout_length` field of `__stream` (in seconds), repeated on each so the timeout survives loss of earlier messages and lets a different executor recover the exchange mid-stream; response-direction messages (executor → invoker) omit it. Seconds align with the MQTT message expiry interval used for other timeouts, keep the header small for long-running streams, and avoid implying a sub-second precision that isn't meaningful.

A local timeout terminates only that side's exchange state; neither side sends a timeout status to the other. The peer reaches its own timeout independently if no further progress occurs.

##### Message level timeout

> We will allow users to set the message expiry interval of each message in a request/response stream; by default it equals the exchange timeout. Every stream message _must_ include a positive, finite message expiry — a message with no (or zero) expiry is rejected. The receiving end uses this value as the de-dup cache length for the cached message (vanilla RPC has the [same requirement](../../reference/command-timeouts.md#input-values)).

Reviewer 3: will we allow a message expiry interval > remaining exchange level timeout? (I think we probably shouldn't allow this)

> We will allow users to set the message expiry interval of each message in a request/response stream; by default it equals the exchange timeout. Every stream message _must_ include a positive, finite message expiry — a message with no (or zero) expiry is rejected. The receiving end uses this value as the de-dup cache length for the cached message (vanilla RPC has the [same requirement](../../reference/command-timeouts.md#input-values)).

Reviewer 3: just to clarify, it would be the remaining exchange timeout, not the entire original exchange timeout, right?
```suggestion
We will allow users to set the message expiry interval of each message in a request/response stream; by default it equals the remaining exchange timeout. Every stream message _must_ include a positive, finite message expiry — a message with no (or zero) expiry is rejected. The receiving end uses this value as the de-dup cache length for the cached message (vanilla RPC has the [same requirement](../../reference/command-timeouts.md#input-values)).
```

> We will allow users to set the message expiry interval of each message in a request/response stream; by default it equals the exchange timeout. Every stream message _must_ include a positive, finite message expiry — a message with no (or zero) expiry is rejected. The receiving end uses this value as the de-dup cache length for the cached message (vanilla RPC has the [same requirement](../../reference/command-timeouts.md#input-values)).

Reviewer 3: one thing about the de-dup cache - this message expiry interval can be used for removing individual indexes from the de-dup cache, but the overall correlation id tracking (tombstone?) would live longer, right?

Cancellation is **idempotent**: the sender may issue `cancel` more than once while exchange is active. Receiving `Canceled` confirms cancellation.

#### Receiving a cancellation

A receiver replies with a **`Canceled`** [status message](#streaming-user-property) — no payload — carrying:

- `__stream`: `s:<cancel request's index>` (status form referencing the received `cancel`)
- `__stat`: `499` (`Canceled`), plus an optional human-readable `__stMsg`
- `__apErr`: `false` (cancellation is not an application error)

An **invoker** sending this to acknowledge an executor-initiated cancellation publishes on the command topic (request direction), so its `__stream` takes the `s:<index>:<timeout>` form — also carrying the remaining timeout; all else is identical.

Whether a receiver replies depends on its state:

- **Still active** — notifies the application, replies with `Canceled`, stops production of the outbound stream, transitions to the canceled state.

> - **Already canceled** — re-sends `Canceled` so a later (re-issued) cancellation is answered.

Reviewer 3: Is there any real difference between both sides sending `c:...:cancel` and one side sending `c:...:cancel` and the other responding with `s:<index>:...` with the `__stMsg`? I'm wondering if we really need both, or if we could just use one or the other to communicate about cancellation

- **Other terminal states** — acknowledges the message and sends nothing.

### Error handling

> Either party may report a problem with an **individual message it received** by sending a [status message](#streaming-user-property) back to that message's sender: `s:<index>` references the offending message and `__stat` carries the error code (with an optional human-readable `__stMsg`). `__apErr` marks whether it is a framework/protocol error (`false`) or an application-level error (`true`) the receiving side's application deliberately returned.

Reviewer 3: Returning __apErr:false things will be easy for this, but the surface to return __apErr:true is interesting - how will we do this (or is this always inferred? Can it be inferred with streaming?)? Since unlike vanilla RPC, the customer doesn't _have_ to respond to every message.
Additionally, with the other ways we recommend customers to report errors in vanilla rpc (outside of explicit SDK support), I would assume that these application desired errors would be sent as data `d:...` messages rather than `s:...` messages and they would be responsible for including the erroneous index in the payload bits if they want rather than us getting involved (there's some weirdness to this plan, but I feel strongly against the alternative)

> Either party may report a problem with an **individual message it received** by sending a [status message](#streaming-user-property) back to that message's sender: `s:<index>` references the offending message and `__stat` carries the error code (with an optional human-readable `__stMsg`). `__apErr` marks whether it is a framework/protocol error (`false`) or an application-level error (`true`) the receiving side's application deliberately returned.

Reviewer 2: I'm not sure we want to build in support for signaling application-level errors like this. With vanilla RPC, we just offered a couple helper methods that inserted a header into/read a header from an RPC message. See [here](https://github.com/Azure/iot-operations-sdks/blob/maxim/streaming-adr-2/doc/dev/adr/0021-error-modeling-headers.md)

Reviewer 2: I think we can punt on adding that same support to RPC streaming if it is requested later. For now, I would just focus on the design for signaling protocol-level errors

> Either party may report a problem with an **individual message it received** by sending a [status message](#streaming-user-property) back to that message's sender: `s:<index>` references the offending message and `__stat` carries the error code (with an optional human-readable `__stMsg`). `__apErr` marks whether it is a framework/protocol error (`false`) or an application-level error (`true`) the receiving side's application deliberately returned.

Reviewer 2: As discussed, we want to avoid trying to make a way to communicate app-level errors like initially proposed here.

Reviewer 2: Additionally, as discussed, we want to make actual protocol level errors (caused by messages that belong to the stream b/c the correlation Id matches) sever the exchange on both sides by having the recipient of the protocol violating message send back an error message to the sender. Both parties understand at that point that the exchange is over and that no further stream entries should be sent

> A per-message error is **not** terminal — neither stream closes and both parties proceed. The original sender receives the status and surfaces it to its application, associated with the referenced entry (for example, to flag a rejected item in a batch). Accepted messages carry no status; success is implicit.

Reviewer 3: > The original sender receives the status and surfaces it to its application, associated with the referenced entry (for example, to flag a rejected item in a batch)

Is this something that we'll be able to do (correlate an error status response to a specific request/response)? Because requests/responses aren't 1:1, I'm not sure how long we'd wait in the SDKs to say that something did or didn't get an error response before returning Ok/Err to the SDK about it. If this sentence is just saying that we'll have a way to surface errors to the user and include the index with the error for context though, that seems fine to me. I'm just not sure we'd be able to programmatically link the error to the message for them

### Exchange termination

An exchange terminates in exactly one of three ways:

- **Graceful** — each producer closes its own stream with `isLast`; the exchange completes once both streams have closed (see [exchange completion](#exchange-completion)).
- **Cancellation** — either side cancels and the `Canceled` (`499`) terminal ends the exchange (see [cancellation support](#cancellation-support)).
- **Timeout** — the [exchange timeout](#exchange-level-timeout) fires.

There is no terminal *error* status: a fatal failure — a crashed peer, an unhandled exception, or a request pump that throws — surfaces as a best-effort **cancellation**, or, failing that, as the peer's timeout.

### Disconnection and recovery

> Streaming inherits MQTT 5's reconnection semantics; recovery hinges on whether the disconnected side's **session** survives. With a persistent session (clean start off, within its session-expiry interval), queued outbound PUBLISHes flush and unacknowledged inbound PUBLISHes redeliver at QoS 1 — de-duplicated by index — so the exchange resumes where it left off. If the session is lost, that side's stream state is gone, the exchange cannot resume, and the peer falls back to its [exchange timeout](#exchange-level-timeout).

Reviewer: Subtle Nit: this is not the MQTT5 reconnection/recovery semantics, it's the `azure_iot_operations_mqtt` Session reconnection/recovery semantics, which are more restrictive.

An in-flight message survives at the broker only for its [message expiry](#message-level-timeout), which the broker **decrements by the time the message sat queued** — so a redelivered entry has less remaining expiry than when sent and may lapse before reconnection (that entry is then lost, which is tolerated since entries are self-contained). A message's remaining expiry therefore does **not** track the exchange-timeout countdown; the two are independent clocks.

> Because the command topic is a [shared subscription](#topics-and-routing), an executor crash does not strand the exchange — another executor in the group takes over, recovering mid-stream from the correlation, the repeated stream metadata, and each request's `timeout_length`. The **invoker has no equivalent**: its response topic is unique to it, so if the invoker's session is lost, in-flight responses cannot be re-routed and are **lost with no recovery** (there is no invoker-side load-balancing). This is an accepted limitation.

Reviewer 3: if the executor crashes and the same or a new executor without context then picks up the stream, the response stream index would then start again at 1, right? This will cause major dedup issues on the invoker side - any indexes already used will be ignored even though they have new contents.

Reviewer 2: To me, this signals that the invoker side's de-dup cache should also be tracking the executor id that the response stream entry came from.

Reviewer 3: what if the same executor restarts and picks it up again (without context)

> If the invoker sends the first request with **no executor subscribed**, the broker returns a `no matching subscribers` PUBACK and drops the request. This is **retriable** — the executor is simply not present yet, not a terminal error — so the invoker re-publishes until an executor appears or the exchange times out. Implementations must treat this reason code uniformly.

Reviewer: Do we know for sure the broker does this? I could be wrong, but I don't recall the broker implementing this behavior.

I might add that I believe this is also incongruent with the existing RPC where no matching subscribers is treated as a success (which, at an MQTT level, it is).

I do not believe there is any high-level retry like this in either Telemetry or RPC, and I'm inclined to not want to get into that business since we start making the protocol stateful. Historically, attempts to get into this level of retry are error-prone.

Reviewer 2: MQ does implement "NoMatchingSubscribers" as a PUBACK return code

> If the invoker sends the first request with **no executor subscribed**, the broker returns a `no matching subscribers` PUBACK and drops the request. This is **retriable** — the executor is simply not present yet, not a terminal error — so the invoker re-publishes until an executor appears or the exchange times out. Implementations must treat this reason code uniformly.

Reviewer 3: > Implementations must treat this reason code uniformly.

this will be an issue - the rust sdk doesn't treat this puback rc as an error.

I also don't think the SDK should handle any retries in this scenario - the application can if they'd like. Otherwise with long exchange timeouts, this could be retrying for effectively forever.

> If the invoker sends the first request with **no executor subscribed**, the broker returns a `no matching subscribers` PUBACK and drops the request. This is **retriable** — the executor is simply not present yet, not a terminal error — so the invoker re-publishes until an executor appears or the exchange times out. Implementations must treat this reason code uniformly.

Reviewer 3: side question - do we want the same behavior for this case whether it's the first request or a later one?

Author: same behavior

> If the invoker publishes the first request with **no executor subscribed**, the broker returns `no matching subscribers` and the request reaches no one. This is an MQTT-level success rather than a terminal exchange error.An invoker that finds no executor either surfaces the reason code or waits out its [exchange timeout](#exchange-level-timeout).

Reviewer: There's also a potential success path here - the executor connects, and the broker eventually delivers it assuming it did not expire.

Again, you'd have to check that the MQ broker supports this (it may or may not), but at an MQTT level there are theoretical good outcomes per the spec - that's why it's a success.

Reviewer 2: As discussed, we are fine with assuming the broker will not attempt to deliver the PUBLISH even if a matching subscriber comes online shortly thereafter. Maybe add a note about this in the appendix.

In the event that an invoker gets this "NoMatchingSubscribers" PUBACK reason, the invoker should throw some "NoAvailableStreamingExecutor" error to signal to the user that the stream cannot push any messages right now.

# ADR 25: RPC Streaming

## Context

Users have expressed a desire to allow more than one request and/or more than one response per RPC invocation.

> ## Requirements

Author: state that request stream can be picked up by another or same when executor crashes for any reason.

   - A producer MUST be able to send its first request or response without knowing the total number of entries in advance.
   - A stream MAY contain exactly one entry.
 - When exposed to the user, each request and response MUST include the index of its position within the stream.
 - The protocol MUST support multiple independent streaming commands simultaneously.
 - The invoker and the executor MUST each be able to cancel the streamed request and/or the streamed response at any time.
 - The invoker and the executor MUST be able to send their requests and responses at arbitrary times relative to one another, except that the invoker MUST initiate the streaming invocation with a request.
   - For instance, the executor MAY send a response as soon as it receives a request, or it MAY wait until the request stream ends before sending its first response.
   - Likewise, the invoker MAY send a request in reaction to a received response.
 - The invoker and the executor MUST each be able to end their own stream gracefully at any time.
   - A side that does not know in advance whether an entry will be its last MUST still be able to end its stream afterward without sending another full entry.

> ## Non-requirements

Author: We don't want/can have resumable invoker

# ADR 25: RPC Streaming

## Context

Users have expressed a desire to allow more than one request and/or more than one response per RPC invocation.

## Requirements

 - A single command invocation MUST support an arbitrary number of command requests and responses.
   - A producer MUST be able to send its first request or response without knowing the total number of entries in advance.
   - A stream MAY contain exactly one entry.

>  - When exposed to the user, each request and response MUST include the index of its position within the stream.

Reviewer 4: Why is this a requirement?

>  - The protocol MUST support multiple independent streaming commands simultaneously.

Reviewer 4: I think this is a SHOULD

 - The invoker and the executor MUST each be able to cancel the streamed request and/or the streamed response at any time.
 - The invoker and the executor MUST be able to send their requests and responses at arbitrary times relative to one another, except that the invoker MUST initiate the streaming invocation with a request.
   - For instance, the executor MAY send a response as soon as it receives a request, or it MAY wait until the request stream ends before sending its first response.
   - Likewise, the invoker MAY send a request in reaction to a received response.
 - The invoker and the executor MUST each be able to end their own stream gracefully at any time.
   - A side that does not know in advance whether an entry will be its last MUST still be able to end its stream afterward without sending another full entry.

>  - If an executor crashes for any reason, the request stream MUST be able to be picked up by another executor in the group — or by the same executor after it restarts.

Reviewer 4: What about Error Handling?

## Non-requirements

 - Different payload shapes per command response/request

>  - Chunking a single large payload across multiple stream messages

Reviewer 4: Again, what does mean? Why is this a non-Goal?

   - Any stream message may be lost (message expiry or other circumstances), so a chunked payload could not be reliably reassembled.
 - Delivering streamed requests/responses in their **intended** order.

>    - Any stream message may be lost, so entries are surfaced in the order received; reconstructing the intended order is left to the application.

Reviewer 4: In what scenario would a message get lost with QoS1?

 - Resuming an exchange after the **invoker** crashes or loses its session.
   - The invoker's response topic is unique to it (no load-balancing), so in-flight responses cannot be re-routed.

## State of the art

gRPC supports these patterns for RPC:
- [Unary RPC](https://grpc.io/docs/what-is-grpc/core-concepts/#unary-rpc) (1 request message, 1 response message)
- [Server streaming RPC](https://grpc.io/docs/what-is-grpc/core-concepts/#server-streaming-rpc) (1 request message, many response messages)
- [Client streaming RPC](https://grpc.io/docs/what-is-grpc/core-concepts/#client-streaming-rpc) (many request messages, one response message)
- [Bi-directional streaming RPC](https://grpc.io/docs/what-is-grpc/core-concepts/#bidirectional-streaming-rpc) (many request messages, many response messages. Request and response stream may send concurrently and/or in any order)

[gRPC also allows for either the client or server to cancel an RPC at any time](https://grpc.io/docs/what-is-grpc/core-concepts/#cancelling-an-rpc)

## Decision

> ### Conceptual model

Reviewer 4: Can we add sequence diagrams to explain the flows better? See the rpc document for examples? 
The sequence diagrams should also cover boundary conditions.

> ### Conceptual model

Reviewer 4: How does the WoT look like with streaming support?

The model is defined here language-agnostically; the [appendix](#illustrative-net-api) gives a concrete C# sketch.

While RPC streaming shares a lot with normal RPC, we define a new communication pattern with two roles — a **streaming command invoker** and a **streaming command executor** — analogous to the existing command invoker and executor.

#### Stream entries and metadata scopes

Each entry in a request or response stream pairs a user **payload** with **metadata**. That metadata combines two scopes:

> - **Message metadata** is scoped to the individual stream entry it travels with.

Reviewer 4: What does this include? 
What user properties are these? What values do they have?

| `control_command_word` | `last` | **control** form only — `last` is the standalone final message that closes the producer's stream (no payload or application-provided user properties). |
| `timeout_length` | uint | The invoker's current timeout counter (exchange time remaining), in seconds; **request-direction only** (invoker → executor), omitted on response-direction messages. |

Examples:

- ```d:0:10``` — a request-direction data message at stream index `0`; the invoker's remaining exchange timeout is 10 seconds.
- ```c:4:last:6``` — the request stream's final (`last`) message at index `4`; the invoker's remaining exchange timeout has dropped to 6 seconds (the field is a live countdown).
- ```s:1:10``` — a request-direction status; its `__stat` value carries the outcome — an error code about the received message at index `1`, or `499` (`Canceled`) to cancel the whole exchange.
- ```d:0``` / ```c:7:last``` / ```s:3``` — the response-direction counterparts (executor → invoker); identical forms except `timeout_length` is omitted.

> Every MQTT PUBLISH belonging to a streaming exchange must include `__stream` in exactly one of these three forms: data messages use the `d` form, control messages use the `c` form (`c:…:last`), and status messages use the `s` form. A status message carries no stream entry; its `__stat` property carries the outcome — an error about a received message, or `499` (`Canceled`) for the whole exchange.

Reviewer 4: What are control messages, data messages and status messages? Can you pls add some details?

[see cancellation support](#cancellation-support) and [timeout support](#timeout-support) for how these fields are used.

#### Topics and routing

A single **correlation GUID** identifies the whole exchange; every message of both streams carries it. The exchange uses **two MQTT topics** — the **request topic** (invoker → executor) and the **response topic** (executor → invoker) — and each carries that direction's data, control, and status messages together.

> The executor subscribes to the **command topic** with a **shared subscription** (so that, with multiple executors, only one handles each exchange), with the same topic pre/suffixing and custom-topic-token support as vanilla RPC. The **response topic** is `clients/{invoker client id}/...` (prefixed like vanilla RPC), unique to the invoker and not shared; the invoker subscribes to it before publishing. Each request-stream message the invoker publishes carries the response topic (so the executor knows where to reply) and a `$partition` user property set to the invoker's client id.

Reviewer 4: How do we make sure the broker sends all messages in one streaming request to the same executor?

Reviewer 4: okay, answered below.

> Because the command topic is a shared subscription, **every** command-topic packet for an exchange — request data, a `last` request, and the invoker's `Canceled` status (whether initiating cancellation or confirming it) — must carry that same `$partition`, so the broker routes them all to the same executor; otherwise a packet could reach a different executor that holds no state for the correlation and be silently dropped. Response-topic packets need only the correlation data, since that topic is already unique to the invoker.

Reviewer 4: But we need to define the $partition value here. What should the invoker use?

#### Exchange lifetime

Because closing one stream does not end the exchange (see [exchange completion](#exchange-completion)), each endpoint keeps its per-correlation state active until the exchange is terminal, so control still flows after a `last`. Once a side is terminal, further data messages for that correlation are acknowledged and ignored; only the required control re-answers (for example, re-sending `Canceled` for a re-issued cancellation) are sent. The per-correlation state is kept as a **tombstone** so late or duplicate packets remain routable and are not treated as a new stream; it is retained at least as long as the longest expiry of any packet that could still arrive, plus some buffer.

#### Producing and consuming

A side **consumes** one stream and **produces** the other. The consuming and producing rules below are identical for both roles; two further rules apply to the **executor** only.

**Consuming a stream:**

> - **De-dup caching.** A consumer de-dups received data messages (QoS 1 may re-deliver) by correlationId + index + timestamp — the index distinguishes duplicates within a stream, while the per-entry HLC timestamp separates a genuine redelivery (same index *and* timestamp) from a producer that restarted and reused an index with a new timestamp (see [disconnection and recovery](#disconnection-and-recovery)). Each cache entry is retained for the duration of its message's expiry interval (see [message level timeout](#message-level-timeout)), even beyond the end of the stream: clearing it when the stream finishes would let a late re-delivery still within its expiry window be treated as new, which is unsafe for non-idempotent commands.

Reviewer 4: - Who adds the timestamp? 
- Why timestamp instead of say, sequence ID?

> - **Stream entry acknowledgement.** Acknowledging an entry is a **stream-level** signal that the consumer is done with it; the framework maps it to the underlying transport acknowledgement, which the consumer never handles directly. By default an entry is acknowledged as soon as it is delivered to the user. **Manual** acknowledgement — holding the ack until processing finishes, so an unacknowledged entry is redelivered rather than lost if the consumer crashes first — is available only to the **executor** (request consumption); the invoker cannot resume a response stream after a crash, so it always auto-acknowledges.

Reviewer 4: What happens when the executor waits beyond the MQTT Broker receive maximum value?

- **`last` receipt.** On a `last` control message (`c:…:last`), the consumer notifies the user that the stream has ended; it is **not** surfaced as a stream entry ([why `last` is its own message](#last-message-being-its-own-message)). Because delivery order is guaranteed, receiving further data for that stream after its `last` is a protocol violation.

> **Producing a stream:** every data message carries the same correlation data, the appropriate [`__stream` metadata](#streaming-user-property), the serialized user payload, and any message metadata plus the stream metadata, at QoS 1. The producer ends its stream with a standalone `last` message on the same topic and correlation. Which topic each side uses, and the `$partition` requirement on the command topic, are covered in [topics and routing](#topics-and-routing) above.

Reviewer 4: Can you please add some details on what the message headers look like for the invocation? 
Instead of words, actual snippets with the headers will help clarify.

A **protocol violation** is a message that **belongs to the exchange** (its correlationId matches) yet breaks the wire contract. Concretely, a correlation-matched message violates the protocol when it:

- carries a payload that cannot be **deserialized** to the stream's expected type;
- has a missing or **malformed `__stream`** property;
- declares an **incompatible streaming protocol version**;
- is published at **QoS 0** (every stream message must be QoS 1); or
- **breaks stream sequencing** — a data entry after the stream's `last`, or a request stream whose `last` arrives with no preceding data entry.

Any such violation is **terminal**: the recipient sends a [status message](#streaming-user-property) back to the sender (`s:<index>` references the offending message, `__stat` carries the error code, with an optional human-readable `__stMsg`), and **both parties then treat the exchange as over and send no further entries**. The referenced index is **diagnostic context only** — since the exchange ends, nothing needs to correlate it back to a specific message. Unmatched or junk data (no correlating exchange) is acknowledged and discarded, never terminating a stream.

> None of these should occur between conforming implementations — they indicate data corruption or a peer that does not follow the protocol. Application-level errors are **out of scope**: the protocol does not carry them, so an application that wants to signal one sends it as an ordinary data entry. The vanilla-RPC `__apErr` (`IsApplicationError`) header is **never set on a streaming message** and is **not** part of the streaming protocol. A streaming message that nonetheless carries `__apErr` is accepted and ignored. Accepted messages carry no status; success is implicit.

Reviewer 4: Why is that? What should an app do if it finds an error in one of the messages it receives? (say the executor)?

### Exchange termination

An exchange terminates in exactly one of four ways:

- **Graceful** — each producer closes its own stream with `last`; the exchange completes once both streams have closed (see [exchange completion](#exchange-completion)).
- **Cancellation** — either side sends a `Canceled` (`499`) status and the exchange ends (see [cancellation support](#cancellation-support)).
- **Timeout** — the [exchange timeout](#exchange-level-timeout) fires.
- **Protocol violation** — the recipient of a malformed in-exchange message returns an error status and both sides end the exchange (see [error handling](#error-handling)).

A fatal failure — a crashed peer, an unhandled exception, or a request pump that throws — surfaces as a best-effort **cancellation**, or, failing that, as the peer's timeout.

> ### Disconnection and recovery

Reviewer 4: Can we split it into the boundary conditions: 
- Invoker connection drops and reconnects same session
- Executor connection drop mid stream and reconnects same session
- Invoker crashes and loses session
- Invoker crashes and reconnects with same session, but loses state
- Executor crashes and loses session
- Executor crashes and reconnects with same session, but loses state

# ADR 25: RPC Streaming

## Context

> Users have expressed a desire to allow more than one request and/or more than one response per RPC invocation.

Reviewer 4: This is not complete or entirely accurate. This seems to indicate that the scenario is where the invoker sends multiple individual requests in one go and expects multiple individual responses in return. I have not heard of this scenario, and should be possible without Streaming support. 
The primary scenario is where the request or response payloads cannot fit in a single MQTT message. 
Please add some more context to this section

Reviewer: Is that not the previously discussed "chunking" scenario? If that's what we want for streaming, does this mean a single message being lost/corrupted invalidates the entire thing? And we can assume there's a definitive end to a stream, rather than being indefinite?

Reviewer 2: +1 to Carter's question here. My understanding is that "streaming" follows what gRPC defines [here](https://grpc.io/docs/what-is-grpc/core-concepts/#bidirectional-streaming-rpc)

Reviewer 4: Well, yes, but that is not what users have asked for. The ask has come from scenarios we have uncovered where we need to send larger payloads on top of MQTT. 
Chunking is the next step of streaming, where the SDK takes care of breaking up and putting together the large payloads. In Streaming, the app can do that by themselves. 
Tke the case of OPC-UA discovering many assets, and returning them at the same time. They can send each asset as a streamed response.

> ## Requirements

Reviewer 4: Pls write requirements in [RFC2119 format](https://datatracker.ietf.org/doc/html/rfc2119)

 - Allow for an arbitrary number of command requests and responses for a single command invocation
   - The total number of requests and responses does not need to be known before the first request/response is sent
   - The total number of entries in a stream is allowed to be 1

>  - When exposed to the user, each request and response includes an index of where it was in the stream

Reviewer 4: What does this mean?

Author: user knows place of the message in the stream.

 - Allow for multiple separate commands to be streamed simultaneously

>  - Allow for invoker and/or executor to cancel a streamed request and/or streamed response at any time

Reviewer: What should happen with packets already in-flight when this happens?

 - Allow for invoker + executor to send their requests/responses at arbitrary* times
   - For instance, executor may send 1 response upon receiving 1 request (stream message), or it may wait for the request stream to finish before sending the first response
   - Alternatively, this allows the invoker to send a request upon receiving a response
   - *The only limitation is that the invoker must initiate the RPC streaming with a request
 - Allow for invoker/executor to end their own request/response stream gracefully at any time
   - For instance, if the executor doesn't know if a response will be the last one prior to sending it, the executor should still be capable of ending the response stream later without sending another fully-fledged payload

## Non-requirements

>  - Different payload shapes per command response/request

Reviewer 4: What does this mean?

Author: I understand it as payload shape is not changing midstream. @timtay-microsoft, please confirm.

Reviewer 2: All message payloads in a stream should be deserializable as the same type by the receiver

Reviewer 4: Got it, pls reword it then. The way Tim re-words it is better, and can be in a GOAL, not non-GOALs

Author: Will do. Moving to Requirements as:

``- All entries in a given stream MUST be deserializable as the same type by the consumer; the payload shape MUST NOT change mid-stream``

   - This may be added as a feature later if requested by customers

## State of the art

gRPC supports these patterns for RPC:
- [Unary RPC](https://grpc.io/docs/what-is-grpc/core-concepts/#unary-rpc) (1 request message, 1 response message)
- [Server streaming RPC](https://grpc.io/docs/what-is-grpc/core-concepts/#server-streaming-rpc) (1 request message, many response messages)
- [Client streaming RPC](https://grpc.io/docs/what-is-grpc/core-concepts/#client-streaming-rpc) (many request messages, one response message)
- [Bi-directional streaming RPC](https://grpc.io/docs/what-is-grpc/core-concepts/#bidirectional-streaming-rpc) (many request messages, many response messages. Request and response stream may send concurrently and/or in any order)

[gRPC also allows for either the client or server to cancel an RPC at any time](https://grpc.io/docs/what-is-grpc/core-concepts/#cancelling-an-rpc)

## Decision

### API design

Described here language-agnostically; the [appendix](#illustrative-net-api) gives a concrete C# sketch. The SDKs target Rust, .NET, and Go.

> While RPC streaming shares a lot with normal RPC, we define a new communication pattern with two roles — a **streaming command invoker** and a **streaming command executor** — analogous to the existing command invoker and executor.

Reviewer: Out of curiosity, how does gRPC handle this?

#### Stream entries and metadata scopes

Each entry in a request or response stream pairs a user **payload** with **per-message metadata**. Streaming distinguishes two metadata scopes:

- **Per-message metadata** travels with each individual stream entry.

> - **Per-stream metadata** applies to a whole stream: the producer attaches it and the consumer reads it. The request and response streams carry **different** per-stream metadata (mirroring vanilla RPC's request vs response metadata), so this scope is **asymmetric** across directions. On the wire it repeats on every message (like the [stream-level timeout](#stream-level-timeout)) to survive first-message loss, and is read once.

Reviewer 3: would be good to define in this ADR what all of the metadata is and what goes where

> - **Per-stream metadata** applies to a whole stream: the producer attaches it and the consumer reads it. The request and response streams carry **different** per-stream metadata (mirroring vanilla RPC's request vs response metadata), so this scope is **asymmetric** across directions. On the wire it repeats on every message (like the [stream-level timeout](#stream-level-timeout)) to survive first-message loss, and is read once.

Reviewer: We should probably formally model what kind of loss is and isn't possible under our delivery guarantees so we're all operating on the same understanding of what can happen and not tying ourselves in knots to avoid impossible scenarios.

> - **Per-stream metadata** applies to a whole stream: the producer attaches it and the consumer reads it. The request and response streams carry **different** per-stream metadata (mirroring vanilla RPC's request vs response metadata), so this scope is **asymmetric** across directions. On the wire it repeats on every message (like the [stream-level timeout](#stream-level-timeout)) to survive first-message loss, and is read once.

Reviewer: Could we not just establish the metadata as part of an exchange to open the stream? Then we would guarantee it would be received, because the stream couldn't start until both sides shake. It's an extra back and forth, but it means we wouldn't have to include metadata on every packet thereafter.

#### Core abstractions

> Each side both produces and consumes a stream — the invoker produces requests and consumes responses, the executor does the reverse — and the two together form one **exchange**, the unit of cancellation and timeout.

Reviewer: Does the concept of a request and response really make sense in a stream context? Also, does this imply that every piece of data sent gets a response? That seems like a lot of unnecessary traffic, and at that point, is kind of a layer on top of our existing RPC.

> Two abstractions carry these across the API:

Reviewer: Can these be described more abstractly? This seems to be more implementation detail than concept, and I struggle to understand the conceptual idea behind it, which makes it a bit difficult to imagine translating to, say, a Rust implementation.

#### Invoker side

The invoker supplies the outbound **request stream** (an async sequence of request entries) together with that stream's metadata; it must contain at least one entry. The invocation establishes the exchange; it does **not** represent completion of the request stream. The SDK activates response reception, sends the mandatory first request, and then returns the inbound **response stream** together with the *exchange context* — without waiting for the second request or for the request stream to end. Early responses are buffered for iteration.

After returning, both streams proceed concurrently, so each can react to the other. The response stream exposes response data and metadata; the exchange context exposes lifecycle and control.

> An empty request stream or setup error fails the invocation before an exchange is returned. Any later request-sending error terminates the local exchange, stops request publication, and triggers a best-effort cancellation; it is exposed through the exchange context's completion signal and, while still open, the response stream.

Reviewer 3: is this desirable (ending the entire exchange because one request fails)? I would say this probably should just fail that single message

> An empty request stream or setup error fails the invocation before an exchange is returned. Any later request-sending error terminates the local exchange, stops request publication, and triggers a best-effort cancellation; it is exposed through the exchange context's completion signal and, while still open, the response stream.

Reviewer: Any single failure killing the stream might be a bit much depending on what's being transmitted.

#### Executor side

> The streaming command executor's callback notifies the user that a command was received; it takes the inbound **request stream** (a *stream context*), that request stream's metadata, and the *exchange context*, and returns the outbound **response stream** (an async sequence of response entries) together with that stream's metadata.

Reviewer: Again, this should be more abstract and divorced from implementation. The Rust SDK does not have callbacks.

> With this design, commands that use streaming are defined at codegen time. Codegen layer changes will be defined in a separate ADR, though.

Reviewer 2: We may want to omit any mention of codegen in this ADR. I don't know that we have or will ever commit to modeling RPC streaming in WoT

Reviewer: Agree

### MQTT layer protocol

#### Streaming user property

To convey streaming context, each message carries a `__stream` MQTT user property with the value:

```<index>:<isLast>:<cancelRequest>:<stream timeout seconds>```

with data types

```<uint>:<boolean>:<boolean>:<uint>```

> The `:<stream timeout seconds>` field is present only in request-stream messages. It carries the effective **idle timeout** — the inactivity window after which a stalled exchange times out (the user-supplied value, or 10 seconds by default).

Reviewer 3: is this a change in meaning? I thought previously that the stream timeout was a timeout over the _entire_ stream. I'm not opposed to this idle timeout though, it seems more useful than trying to guess how long an entire stream would be active for

**Table 1. `__stream` fields, in the order they appear in the value.**

| Field | Type | Present in | Meaning | Ignored when |
|-------|------|------------|---------|--------------|
| `index` | uint | every stream message | Position of this message within the stream | `cancelRequest = true` or a terminal error status |
| `isLast` | bool | every stream message | `true` on the standalone final message that closes the stream (no payload or application-provided user properties) | `cancelRequest = true` or a terminal error status |

> | `cancelRequest` | bool | every stream message | `true` marks a cancellation request for the stream | — (always meaningful) |

Reviewer: Can there be data AND a cancellation request?

| `stream timeout s` | uint | **request** stream messages only | Effective idle-timeout window in seconds | Never present on response messages |

Examples:

- ```0:false:false:10``` — first (not last) request message, 10-second idle timeout.
- ```3:true:false``` — the final response message.
- ```0:true:true``` — a cancellation; `index` and `isLast` are ignored (the request-direction form additionally carries the timeout field, e.g. ```0:true:true:10```).

> Every MQTT PUBLISH belonging to a streaming exchange, **including a terminal status**, must include `__stream`. A terminal status uses `index = 0`, `isLast = false`, and `cancelRequest = false`; its `__stat` value terminates the exchange. The request-direction form also carries the effective idle timeout; the response-direction form has only the first three fields.

Reviewer: This contradicts rather than clarifies - if only the request side has the full set of property values, then you shouldn't state that all messages have them on line 83 above, you should be clear up front that request and response have different values.

[see cancellation support](#cancellation-support) and [timeout support](#timeout-support) for how these fields are used.

#### Exchange routing and lifetime

> Each of the two MQTT topics is a **directional route** that multiplexes two logical lanes for a correlation: a **data lane** (stream entries) and an **exchange-control lane** (`isLast`, cancellation, `Canceled`, and error statuses). An `isLast` message closes only the **data** lane in its direction; it does **not** tear down the route. Both routes — the topic subscription and the per-correlation exchange state behind it — stay active until the whole **exchange** reaches a terminal state, so cancellation and terminal-status controls can still flow after a half-stream's `isLast`.

Reviewer 2: This document is a bit confusing/vague about what MQTT topics it is referring to here. Is it talking about the request and response topics? If so, please be explicit about that

Reviewer 2: I'm also not following this concept of "two logical lanes for a correlation". Are you suggesting that stream entries vs control messages are sent on different topics or something?

Reviewer 3: +1

Author: > I'm also not following this concept of "two logical lanes for a correlation". Are you suggesting that stream entries vs control messages are sent on different topics or something?

no - single topic for data and control in one direction (logical lanes), when data lane "closed" (no more payload carrying stream entries delivered) - control lane stays "open" until terminal state of the whole exchange.

Reviewer: I am also highly confused by this, especially given that the data in the "exchange control lane" seems to have been previously described as being provided via user properties on the data messages.

> Each of the two MQTT topics is a **directional route** that multiplexes two logical lanes for a correlation: a **data lane** (stream entries) and an **exchange-control lane** (`isLast`, cancellation, `Canceled`, and error statuses). An `isLast` message closes only the **data** lane in its direction; it does **not** tear down the route. Both routes — the topic subscription and the per-correlation exchange state behind it — stay active until the whole **exchange** reaches a terminal state, so cancellation and terminal-status controls can still flow after a half-stream's `isLast`.

Reviewer 3: > Both routes — the topic subscription and the per-correlation exchange state behind it — stay active until the whole **exchange** reaches a terminal state,

Note that (assuming this is similar to regular rpc), the topic subscription always lasts through the entire lifetime of the invoker/executor because wildcards will be used for any portions that can vary per invocation.

> Because the executor subscribes to the command topic with a shared subscription, **every** command-topic packet for an exchange — request data, an `isLast` request, invoker cancellation, and the invoker's `Canceled` acknowledgement — must carry the same `$partition` value (the invoker's client id). Otherwise the broker may route a control packet to a different executor that holds no state for the correlation, silently dropping it from the exchange. Response-topic packets need only the correlation data, because `clients/{invokerId}/...` is unique to the invoker and is not a shared subscription.

Reviewer 2: I recall this scenario (executor dies mid-stream, invoker's stream entries start being funneled to a new executor) being a loose end in the previous discussions on this feature. It appears that this version of the doc doesn't really go into depth on this scenario either.

I'm not sure that we would want to automatically make this an error scenario for the user. There are probably valid use cases where an invoker is streaming requests to an executor and the other executors are capable of taking over the stream of requests even from the middle. Scenarios such as where the executors are writing the request stream payloads to some shared storage and will return a basic "OK" response once the request stream ends.

I also see scenarios where it does not make sense to try to have another executor pick up the stream of requests mid-stream (such as executors are running on different machines/don't have any shared context).

Since this scenario seems very dependent on user application context, I'd consider adding to the SDK API design a flag that allows users to opt-in (or opt out?) to having the executor return a specific error code upon receiving a streaming request that is not the initial request in a stream.

Reviewer 3: > Since this scenario seems very dependent on user application context, I'd consider adding to the SDK API design a flag that allows users to opt-in (or opt out?) to having the executor return a specific error code upon receiving a streaming request that is not the initial request in a stream.

I'm not sure returning an error is something we can/want to do, because we need to be able to handle not receiving the first streaming request anyways (this is why we have the metadata on all messages, not just the first one) or receiving it out of order. I'm not sure we'd be able to distinguish a request message that was previously handled by a different executor vs a request message where we lost the initial request message

How does the broker currently handle partitions if the original client receiving messages for one partition disappears? I imagine whatever this behavior is is what we have to work with. My guess is that it would only send something to a different executor if the first is no longer available, so there wouldn't be much of a risk of the original executor "missing" a control packet

> Because the executor subscribes to the command topic with a shared subscription, **every** command-topic packet for an exchange — request data, an `isLast` request, invoker cancellation, and the invoker's `Canceled` acknowledgement — must carry the same `$partition` value (the invoker's client id). Otherwise the broker may route a control packet to a different executor that holds no state for the correlation, silently dropping it from the exchange. Response-topic packets need only the correlation data, because `clients/{invokerId}/...` is unique to the invoker and is not a shared subscription.

Reviewer 3: > Response-topic packets need only the correlation data, because `clients/{invokerId}/...` is unique to the invoker and is not a shared subscription.

Note that the topic pattern having `clients/{invokerId}` is just a recommended pattern, it is not required. That said, as you note, invokers don't use a shared subscription, so if there is any response topic pattern overlap, both invokers would receive the message, and they would determine whether they should do anything with it or not based on the correlation id

> Once a side has reached a terminal state, further data messages for that correlation are acknowledged and ignored; only the required control re-answers (for example, re-sending `Canceled` for a retried cancellation) are sent. The per-correlation exchange state is kept as a tombstone so that late or duplicate packets remain routable and are not treated as a new stream; see [stream level timeout](#stream-level-timeout) for how long.

Reviewer 3: > Once a side has reached a terminal state, further data messages for that correlation are acknowledged and ignored;

Do you mean that if the invoker sends request::isLast that it will not give any further responses to the user and if the executor sends response::isLast, it won't give any further requests to the user? Imo I don't think we should be ignoring those.

Or do you mean that if the executor receives request::isLast, then it won't give any further requests received to the user and if the invoker receives response::isLast that it don't give any further responses to the user? This makes a bit more sense to me, although it is unfortunate to have the possibility of some missed messages if they get received out of order (but I don't think there's a way around it)

#### Common stream handling

A side **consumes** one stream and **produces** the other. The rules below are identical for both roles.

**Consuming a stream:**

> - **De-dup caching.** A consumer de-dups received data messages (QoS 1 may re-deliver) by correlationId + index — the index distinguishes duplicates since the correlationId is shared by the whole stream. Each cache entry is retained for the duration of its message's expiry interval (see [message level timeout](#message-level-timeout)), even beyond the end of the stream: clearing it when the stream finishes would let a late re-delivery still within its expiry window be treated as new, which is unsafe for non-idempotent commands.

Reviewer 2: ~~I'm not sure how feasible it is to offer caching support for RPC streaming. Conceptually, streams are simply larger and less bounded than single RPC messages. I would consider reading up on if gRPC even offers this. If they don't, then I think we could get away with not supporting it here.~~

Just understood that you are talking about just de-dup caching, not replay caching. Nevermind on this point. It does seem reasonable to just track which received MQTT messages in a stream have already been given to the reader of the stream

#### Invoker side

The invoker first subscribes to its response topic (`clients/{mqtt client id of invoker}/...`, prefixed like vanilla RPC), then publishes the request stream. In addition to the common producer fields, each request-stream message (data and the closing `isLast`) carries:

- the response topic, so the executor knows where to reply;
- the `$partition` user property set to the invoker's client id, so the shared subscription routes every message of the exchange to the same executor.

#### Executor side

The executor subscribes to the command topic using a **shared subscription** (so that, with multiple executors, only one handles each exchange), with the same topic pre/suffixing and custom-topic-token support as vanilla RPC. On the first request message it notifies the application, which then supplies the response stream; each response message is published to the response topic named in the request, with the same correlation data.

Two executor-only rules:

> - If an `isLast` arrives before any data message in the request stream, log an error, acknowledge it, and ignore it — a request stream must have at least one entry.

Reviewer 3: maybe this is detailed further down, but in the event that the isLast message comes in before the 1st data message in the request stream (which would then open a new exchange), is it a problem that the stream won't ever receive the `isLast` message? I suppose the timeout should eventually handle this scenario?

- Unlike vanilla RPC, the executor keeps **no replay cache**: streams may grow without bound, so replaying a response stream isn't feasible.

### Timeout support

Timeout support avoids either side getting stuck — waiting for a final message that was lost or never sent, or for a peer that has silently stalled (the invoker waiting on responses, or the executor waiting on requests).

#### Decision

> The invoker configures an **idle (inactivity) timeout** for the exchange and a per-message expiry for request/response data. If the user does not specify an idle timeout, the SDK defaults to 10 seconds; a user-supplied value must be positive and finite and is rounded up to whole seconds. Every exchange therefore has a positive, finite effective idle timeout of at least one second.

Reviewer 3: is there a max that we will enforce for the timeout value? Just wondering if having a user set some years-long value for this would cause any worse issues than if we supported no timeout

> The idle timeout measures time **without progress**, not total duration: it is the backstop for an exchange that stalls before reaching [graceful completion](#core-abstractions) — a half-stream that never closes, a lost final message, or a crashed peer. A healthy stream keeps resetting the timer, so it runs for as long as it keeps making progress; this is what lets an exchange stay open indefinitely — for example, a continuous sensor feed that ends only on cancellation or failure. A stream that can fall silent without ending must emit periodic heartbeats to stay within the idle window, or use a correspondingly larger timeout.

Reviewer 2: "without progress" is a bit vague here. I'd consider explicitly stating that the Idle timeout countdown resets each time a stream message is received.

Reviewer 2: Is an invoker going to timeout if the intended flow is that the invoker sends 10,000 messages to the executor, and then the executor just sends an "OK" at the end? Or does sending a message count as "making progress"?

Reviewer: I read this as being basically application-level (from an MQTT perspective) ping, that functions similarly to MQTT ping (sent in lieu of real messages, unnecessary otherwise).

EDIT: later on it says that they occur regularly. I disagree with that choice - I think heartbeat is only necessary if no messages are being exchanged otherwise.

> The idle timeout measures time **without progress**, not total duration: it is the backstop for an exchange that stalls before reaching [graceful completion](#core-abstractions) — a half-stream that never closes, a lost final message, or a crashed peer. A healthy stream keeps resetting the timer, so it runs for as long as it keeps making progress; this is what lets an exchange stay open indefinitely — for example, a continuous sensor feed that ends only on cancellation or failure. A stream that can fall silent without ending must emit periodic heartbeats to stay within the idle window, or use a correspondingly larger timeout.

Reviewer 2: > A stream that can fall silent without ending must **emit periodic heartbeats** to stay within the idle window, or use a correspondingly larger timeout.

This may actually suggest that we need to add some kind of control message (like "IsLast" message) that is a dedicated heartbeat message. It would be cumbersome to ask users to construct such a message.

Our SDK could internally ingest the heartbeat message to reset the timeout without exposing it as a message in the stream

Reviewer 2: Alternatively, this could be a signal that implementing a constantly-resetting timeout timer is a bit unwieldly, and that we should instead go with a simpler approach. 

We could just start the timeout timer upon the first stream message being delivered/received and never reset it.

Reviewer 3: +1 - since we said above that we aren't supporting variable payload shapes, this sounds like something we should do in the background from both sides

Reviewer 3: although the interesting thing is that this might not capture hangs in the application code, only entire application crashes - for example, if my executor is supposed to send a new image frame every 10 seconds, but something in that logic freezes, then the executor (sdk code) would continue to send heartbeats even though this probably should be a timeout scenario.

Maybe this indicates the need for separate request vs response timeouts (although I don't love the extra complications this idea adds), so that if the expected pattern is one request and then 1000 responses 10 seconds apart, the request timeout could be 1 hour and the response timeout could be 1 minute, with no need for additional heartbeat messages. Of course that doesn't tell the executor to stop sending messages if the invoker crashes, but maybe that isn't a goal

Reviewer 3: another thing that might come into play for some of these scenarios (but still not if the invoker was going to send 1 request per hour) - if one side has sent isLast already, then not receiving anything more from them wouldn't trigger a timeout (?)

Reviewer: +1, I don't think any heartbeats should be reaching the end user.

##### Stream level timeout

Each request-stream message carries the effective idle timeout in the `<stream timeout seconds>` field of `__stream`, sent on **all** request messages in case the first N are lost. Seconds align with the MQTT message expiry interval used for other timeouts, keep the header small for long-running streams, and avoid implying a sub-second precision that isn't meaningful.

> Each side runs its own idle countdown for the exchange and **restarts it on every progress event**: receiving either a new, valid stream PUBLISH for the exchange or the first PUBACK for one of its own stream PUBLISH packets.

Reviewer 2: Considering my points above, I don't think it actually makes sense for an invoker/executor to reset its stream timeout each time it gets a PUBACK for a stream entry it sent. This timeout is to defend against a stalled receiver of the stream, and the PUBACK doesn't really tell us anything about the state of a recipient.

With the current approach, an invoker could stream requests indefinitely to a stalled executor which seems undesirable.

I think it does make more sense to only reset this timeout if the invoker/executor receives a stream message or heartbeat from the other party.

Reviewer: +1, the broker in the middle obfuscates the happenings of the other end, it needs to be explict.

> - The **invoker** starts its timer on the first PUBACK for a request PUBLISH and resets it on each new, valid response PUBLISH and the first PUBACK for each subsequent request PUBLISH. If the timer elapses before the exchange has gracefully completed — it has not both sent its `isLast` request and received the `isLast` response — it reports the timeout to the user and stops sending.

Reviewer 3: > it reports the timeout to the user and stops sending.

since the timeout is to detect that the other side has gone silent, should the invoker also send an isLast message or cancellation or something similar so that if the executor is not dead it knows that the invoker is no longer listening? I suppose it would eventually time out as well after not receiving more requests, but for larger timeouts this might be a while and it could be good to know sooner

> - The **executor** starts its timer on the first new, valid request PUBLISH and resets it on each subsequent new, valid request PUBLISH and the first PUBACK for each response PUBLISH. If the timer elapses before it has both received the `isLast` request and sent the `isLast` response, it reports the timeout to the user callback and asks it to stop.

Reviewer 3: > it reports the timeout to the user callback and asks it to stop.

does this mean the user could still keep sending responses, or do we block that flow after the timeout?

A local idle timeout terminates only that side's exchange state; neither side sends a timeout status to the other. The peer reaches its own timeout independently if no further progress occurs.

Because *received* messages also count as progress, a side that only consumes (or only produces) — such as the consumer of a one-directional sensor feed — stays alive as long as the other side keeps data flowing.

Messages received by either side after it has timed out are acknowledged but otherwise ignored. Each party therefore keeps a per-correlation tombstone for timed-out streams so post-timeout packets aren't treated as a new stream; it is retained at least as long as the longest expiry of any packet that could still arrive.

##### Message level timeout

> Users may set the message expiry interval of each **data** message in a request/response stream; by default it equals the effective idle timeout. Every stream data message _must_ include a positive, finite message expiry — a message with no (or zero) expiry is rejected, as in vanilla RPC, since an unbounded expiry would make the de-dup cache grow without bound. The receiving end uses this value as the de-dup cache length for the cached message (vanilla RPC has the [same requirement](../../reference/command-timeouts.md#input-values)).

Reviewer 3: if the stream level timeout is now an idle timeout instead of the timeout of the whole stream, what is the reason to have the message level timeout be any other value? If it's less than the stream level timeout, then we're having the broker toss a message that the other side is still expected to be listening for. And if it's more than the stream level timeout, then it's likely we're asking the broker to continue to deliver a message that the other side is no longer listening for (and could increase the chances of starting a new exchange that should have gotten deduped instead)

### Protocol layer details

Cancellation acknowledgements reuse the same status mechanism as vanilla RPC: the status travels in the `__stat` MQTT user property (with an optional human-readable `__stMsg`), not a separate acknowledgement packet. Streaming introduces one new status code:

- **`Canceled` = `499`** (mirrors the conventional "Client Closed Request" code). Cancellation is not an application error, so `__apErr` is `false`.

A `Canceled` response from the executor to the invoker looks like this on the wire:

```text
PUBLISH
  topic:                 clients/{invokerId}/...        # the stream's response topic
  qos:                   1
  correlationData:       <same GUID as the stream>
  messageExpiryInterval: <control-message expiry defined above>
  userProperties:

>     __stream:  0:false:false        # streaming terminal status, not a cancellation request

Reviewer 3: should isLast be true if it's responding to a cancel request?

Reviewer 3: or not necessarily because it might send more responses if the invoker sends more requests until the invoker accepts the cancellation?

#### Sending a cancellation

Either side cancels by publishing a message with the [cancel flag set](#streaming-user-property), no payload, the same correlation data, on the topic it uses to reach the other party:

- The **invoker** cancels on the command topic and therefore carries `$partition` (see [exchange routing and lifetime](#exchange-routing-and-lifetime)). It keeps listening on the response topic afterward, since a late successful response or the `Canceled` status may still arrive.
- The **executor** cancels on the invoker's response topic and needs only the correlation data. After sending, it listens on the command topic for the invoker's `Canceled` acknowledgement.

The sender may retransmit the cancellation request while its local exchange remains active. Receiving `Canceled` confirms cancellation; any other terminal outcome ends retransmission without confirming it.

#### Receiving a cancellation

> The receiver of a cancellation:

Reviewer 2: ```suggestion
The receiver of a cancellation responds depending on the state of that receiver:
```

- **Still active** — notifies the application (if the RPC is still running) and replies with `Canceled` on the appropriate route (the invoker's acknowledgement travels on the command topic and carries `$partition`).
- **Already completed** (both halves closed) — acknowledges the message and sends nothing.

> - **Already canceled** — re-sends `Canceled` so a retried/duplicate cancellation is answered.

Reviewer 2: Already canceled by the receiver's application code, or already canceled by the sender's application code?

> Once a side is canceled, any further messages for that correlation are acknowledged but not delivered to the user.

Reviewer 3: Do you mean once cancellation (request or confirmation) has been received, or once cancellation has been requested? I feel like until cancellation (request or confirmation) has been received, any further messages should maybe still be delivered to the user since it isn't considered closed yet (I think?)

// stream's metadata (it faults if the exchange ends before any response).
public interface IResponseStreamContext<T> : IStreamContext<T>
    where T : class
{
    Task<ResponseStreamMetadata> StreamMetadata { get; }
}

public interface IExchangeContext
{
    // Completes on graceful close; faults or cancels on any other terminal.
    Task Completion { get; }

    Task CancelAsync(Dictionary<string, string>? userProperties = null, CancellationToken cancellationToken = default);

    // Fires on peer cancel or timeout; use IsCanceled / HasTimedOut to distinguish.
    CancellationToken CancellationToken { get; }

    bool IsCanceled { get; }

    bool HasTimedOut { get; }

    Dictionary<string, string>? GetCancellationRequestUserProperties();
}
```

The invoker supplies the request stream (and its stream-level metadata) and returns the response stream plus the exchange context:

```csharp
public abstract class StreamingCommandInvoker<TReq, TResp>
    where TReq : class
    where TResp : class
{
    // false -> the caller must ack each response entry via ReceivedStreamingExtendedResponse.AcknowledgeAsync.

>     public bool AutomaticallyAcknowledgeResponses { get; set; } = true;

Reviewer 3: I think for the invoker, there's no reason to have manual acknowledgements - if the invoker crashes and restarts, it no longer has the context to handle any responses from a previous request exchange

      IAsyncEnumerable<StreamingExtendedRequest<TReq>> requests,
      RequestStreamMetadata? streamMetadata = null,
      Dictionary<string, string>? additionalTopicTokenMap = null,
      TimeSpan? idleTimeout = default,
      CancellationToken cancellationToken = default) {...}
}
```

The executor's callback receives the request stream, its stream-level metadata, and the exchange context, and returns the response stream together with its metadata:

```csharp
public abstract class StreamingCommandExecutor<TReq, TResp> : IAsyncDisposable
    where TReq : class
    where TResp : class
{
    public required Func<
        IStreamContext<ReceivedStreamingExtendedRequest<TReq>>,
        RequestStreamMetadata,
        IExchangeContext,
        (IAsyncEnumerable<StreamingExtendedResponse<TResp>> Responses, ResponseStreamMetadata Metadata)> OnStreamingCommandReceived { get; set; }

    // false -> the callback must ack each request entry manually.
    public bool AutomaticallyAcknowledgeRequests { get; set; } = true;
}
```

### IsLast message being its own message

> Three approaches to marking the final message in a stream were considered, and why the first two don't work:

Reviewer 2: ```suggestion
Three approaches to marking the final message in a stream were considered, and this is why the other two approaches don't work:
```

#### Common stream handling

A side **consumes** one stream and **produces** the other. The rules below are identical for both roles.

**Consuming a stream:**

- **De-dup caching.** A consumer de-dups received data messages (QoS 1 may re-deliver) by correlationId + index — the index distinguishes duplicates since the correlationId is shared by the whole stream. Each cache entry is retained for the duration of its message's expiry interval (see [message level timeout](#message-level-timeout)), even beyond the end of the stream: clearing it when the stream finishes would let a late re-delivery still within its expiry window be treated as new, which is unsafe for non-idempotent commands.
- **Acknowledgement.** By default a consumer acknowledges each message as soon as it is delivered to the user. Users may opt into manual acknowledgement to finish processing a message before forgoing broker re-delivery on an unexpected crash. 

> - **`isLast` receipt.** On a message with the `isLast` flag set, the consumer notifies the user that the stream has ended. This standalone message carries no payload or application-provided user properties and is **not** surfaced as a stream entry ([why `isLast` is its own message](#islast-message-being-its-own-message)).

Reviewer: Are we supporting "application-provided user properties" (aka "custom user data") ever? Do they make sense in a streaming context?

#### Invoker side

The invoker first subscribes to its response topic (`clients/{mqtt client id of invoker}/...`, prefixed like vanilla RPC), then publishes the request stream. In addition to the common producer fields, each request-stream message (data and the closing `isLast`) carries:

- the response topic, so the executor knows where to reply;
- the `$partition` user property set to the invoker's client id, so the shared subscription routes every message of the exchange to the same executor.

#### Executor side

The executor subscribes to the command topic using a **shared subscription** (so that, with multiple executors, only one handles each exchange), with the same topic pre/suffixing and custom-topic-token support as vanilla RPC. On the first request message it notifies the application, which then supplies the response stream; each response message is published to the response topic named in the request, with the same correlation data.

Two executor-only rules:

> - If an `isLast` arrives before any data message in the request stream, log an error, acknowledge it, and ignore it — a request stream must have at least one entry. A `heartbeat` that arrives before the first request likewise establishes no exchange and is simply acknowledged and ignored.

Reviewer: Is this even possible?

> - Unlike vanilla RPC, the executor keeps **no replay cache**: streams may grow without bound, so replaying a response stream isn't feasible.

Reviewer: I need a clarification here - there is a degree of replay built into the MQTT protocol, so I don't know what you mean here.

### Timeout support

> Timeout support avoids either side getting stuck — waiting for a final message that was lost or never sent, or for a peer that has silently stalled (the invoker waiting on responses, or the executor waiting on requests).

Reviewer: Is this an API timeout or a protocol timeout? i.e. is this just to cancel the API from hanging, or does it involve sending data?

### Protocol layer details

Cancellation acknowledgements reuse the same status mechanism as vanilla RPC: the status travels in the `__stat` MQTT user property (with an optional human-readable `__stMsg`), not a separate acknowledgement packet. Streaming introduces one new status code:

- **`Canceled` = `499`** (mirrors the conventional "Client Closed Request" code). Cancellation is not an application error, so `__apErr` is `false`.

A `Canceled` response from the executor to the invoker looks like this on the wire:

```text
PUBLISH
  topic:                 clients/{invokerId}/...        # the stream's response topic
  qos:                   1
  correlationData:       <same GUID as the stream>
  messageExpiryInterval: <control-message expiry defined above>
  userProperties:
    __stream:  0:false:false        # streaming terminal status, not a cancellation request
    __stat:    499                  # Canceled
    __stMsg:   "Canceled"           # optional

>     __apErr:   false                # cancellation is not an application error

Reviewer: Does this concept make sense when streaming?

The `__apErr` (`IsApplicationError`) property classifies an error as either a framework/protocol error (`__apErr = false`: canceled, bad request, internal error) or an application-level error (`__apErr = true`) the command logic chose to return. **Either way the error status terminates the stream** — there is no per-message error status that leaves the stream running. An application that needs a per-item outcome while the stream keeps going (for example, a batch where individual items may fail) must encode that in its response payload (`TResp`), not the protocol status — a mid-stream "failed item" is just a normal response whose payload represents the failure.

### Disconnection scenario considerations

In every case, QoS 1 sessions carry queued messages across a reconnection (within each message's expiry), and whichever side stops seeing progress reaches its own local [idle timeout](#stream-level-timeout) independently.

- Invoker side disconnects unexpectedly while sending requests
  - On reconnection, the request messages queued in its session client publish as expected and the exchange resumes

>   - Otherwise the executor stops seeing requests and its idle timeout fires

Reviewer: Worth noting here that even if the stream is timed out from either side, if the messages were queued on the invoker side, they're going out when the connection comes back, even if the stream has timed out. Once owned by the MQTT session, the session is required to redeliver to respect the protocol and deliver.

- Invoker side disconnects unexpectedly while receiving responses

>   - The broker holds each published response for its [message-level timeout](#message-level-timeout) (message expiry interval) and redelivers those still within their expiry on reconnection; those whose expiry lapses first are lost

Reviewer: This seems to be inconsistent with previously stated behavior that a failed message kills the stream. Here we're allowing for messages that expire to be lost, so it would appear here we ARE okay with messages being lost.

How is the executor going to handle this? Again, if streaming something that doesn't require every piece of data, not so big a deal, but if streaming a large piece of data which needs all segments, this is catastrophic.

>   - If the invoker's session is lost, the exchange cannot resume and the executor's idle timeout fires

Reviewer: Implementation specific - this is not what would happen in Rust, it would fail immediately.

- **`last` receipt.** On a `last` control message (`c:…:last`), the consumer notifies the user that the stream has ended; it is **not** surfaced as a stream entry ([why `last` is its own message](#last-message-being-its-own-message)). Because delivery order is guaranteed, receiving further data for that stream after its `last` is a protocol violation.

**Producing a stream:** every data message carries the same correlation data, the appropriate [`__stream` metadata](#streaming-user-property), the serialized user payload, and any message metadata plus the stream metadata, at QoS 1. The producer ends its stream with a standalone `last` message on the same topic and correlation. Which topic each side uses, and the `$partition` requirement on the command topic, are covered in [topics and routing](#topics-and-routing) above.

**Executor-only rules:**

> - A `last` in the request stream with no preceding data entry is a [protocol violation](#error-handling) (a request stream must have at least one entry): the executor returns the error status and the exchange ends.

Reviewer 3: couldn't this happen if the executor handling the stream crashes and a new one picks it up? Although I think this behavior still works in that case

Author: no action

A **protocol violation** is a message that **belongs to the exchange** (its correlationId matches) yet breaks the wire contract. Concretely, a correlation-matched message violates the protocol when it:

- carries a payload that cannot be **deserialized** to the stream's expected type;
- has a missing or **malformed `__stream`** property;
- declares an **incompatible streaming protocol version**;
- is published at **QoS 0** (every stream message must be QoS 1); or
- **breaks stream sequencing** — a data entry after the stream's `last`, or a request stream whose `last` arrives with no preceding data entry.

Any such violation is **terminal**: the recipient sends a [status message](#streaming-user-property) back to the sender (`s:<index>` references the offending message, `__stat` carries the error code, with an optional human-readable `__stMsg`), and **both parties then treat the exchange as over and send no further entries**. The referenced index is **diagnostic context only** — since the exchange ends, nothing needs to correlate it back to a specific message. Unmatched or junk data (no correlating exchange) is acknowledged and discarded, never terminating a stream.

> None of these should occur between conforming implementations — they indicate data corruption or a peer that does not follow the protocol. Application-level errors are **out of scope**: the protocol does not carry them, so an application that wants to signal one sends it as an ordinary data entry. Accepted messages carry no status; success is implicit.

Reviewer 3: might be good to call out that `__apErr` can never be true in a streaming protocol message (maybe we shouldn't even include it as a header?)
